protocol/grpc/write_message.go

```diff
@@ -0,0 +1,208 @@
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/binding/format"
+	"github.com/cloudevents/sdk-go/v2/binding/spec"
+	"github.com/cloudevents/sdk-go/v2/types"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// WritePubMessage fills the provided pubMessage with the message m.
+// Using context you can tweak the encoding processing (more details on binding.Write documentation).
+func WritePubMessage(ctx context.Context, m binding.Message, pbEvt *pb.CloudEvent, transformers ...binding.Transformer) error {
+	structuredWriter := (*pbEventWriter)(pbEvt)
+	binaryWriter := (*pbEventWriter)(pbEvt)
+
+	_, err := binding.Write(
+		ctx,
+		m,
+		structuredWriter,
+		binaryWriter,
+		transformers...,
+	)
+	return err
+}
+
+type pbEventWriter pb.CloudEvent
+
+var (
+	_ binding.StructuredWriter = (*pbEventWriter)(nil)
+	_ binding.BinaryWriter     = (*pbEventWriter)(nil)
+)
+
+func (b *pbEventWriter) SetStructuredEvent(ctx context.Context, f format.Format, event io.Reader) error {
+	if b.Attributes == nil {
+		b.Attributes = make(map[string]*pb.CloudEventAttributeValue)
+	}
+
+	b.Attributes[contenttype], _ = attributeFor(f.MediaType())
+
+	var buf bytes.Buffer
+	_, err := io.Copy(&buf, event)
+	if err != nil {
+		return err
+	}
+
+	// TODO: check the data content type and set the right data format
+	b.Data = &pb.CloudEvent_BinaryData{
+		BinaryData: buf.Bytes(),
+	}
+
+	return nil
+}
+
+func (b *pbEventWriter) Start(ctx context.Context) error {
+	if b.Attributes == nil {
+		b.Attributes = make(map[string]*pb.CloudEventAttributeValue)
+	}
+
+	return nil
+}
+
+func (b *pbEventWriter) End(ctx context.Context) error {
+	return nil
+}
+
+func (b *pbEventWriter) SetData(reader io.Reader) error {
+	buf, ok := reader.(*bytes.Buffer)
+	if !ok {
+		buf = new(bytes.Buffer)
+		_, err := io.Copy(buf, reader)
+		if err != nil {
+			return err
+		}
+	}
+
+	b.Data = &pb.CloudEvent_BinaryData{
+		BinaryData: buf.Bytes(),
+	}
+
+	return nil
+}
+
+func (b *pbEventWriter) SetAttribute(attribute spec.Attribute, value interface{}) error {
+	switch attribute.Kind() {
+	case spec.SpecVersion:
+		val, ok := value.(string)
+		if !ok {
+			return fmt.Errorf("invalid SpecVersion type, expected string got %T", value)
+		}
+		b.SpecVersion = val
+		return nil
+	case spec.ID:
+		val, ok := value.(string)
+		if !ok {
+			return fmt.Errorf("invalid ID type, expected string got %T", value)
+		}
+		b.Id = val
+		return nil
+	case spec.Source:
+		val, ok := value.(string)
+		if !ok {
+			return fmt.Errorf("invalid Source type, expected string got %T", value)
+		}
+		b.Source = val
+		return nil
+	case spec.Type:
+		val, ok := value.(string)
+		if !ok {
+			return fmt.Errorf("invalid Type type, expected string got %T", value)
+		}
+		b.Type = val
+		return nil
+	case spec.DataContentType:
+		if value == nil {
+			delete(b.Attributes, contenttype)
+		}
+		b.Attributes[contenttype], _ = attributeFor(value)
```

Comment:
Why do we delete the item from the map just to then re-add it here with `nil` ? Seems like this should be under an `else` or just remove the if+delete on lines 126-128. Same for below

Reply:
good catch, added `else` to handle non-nil case.